### ISSUE TITLE
Could not find directory of OpenSSL installation on Ubuntu 20.x Github Actions

### ISSUE THREAD

**Issue by DevYukine** [sfackler/rust-openssl#1367]
```
error: failed to run custom build command for `openssl-sys v0.9.58`
Error: failed to run custom build command for `openssl-sys v0.9.58`
Caused by:
  process didn't exit successfully: `/target/release/build/openssl-sys-cd6721960b06b7d7/build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=I686_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR
  I686_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=I686_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR
  I686_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=I686_UNKNOWN_LINUX_GNU_OPENSSL_DIR
  I686_UNKNOWN_LINUX_GNU_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_i686-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_i686_unknown_linux_gnu
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_i686-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_i686_unknown_linux_gnu
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_i686-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_i686_unknown_linux_gnu
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_i686-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_i686_unknown_linux_gnu
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  run pkg_config fail: "`\"pkg-config\" \"--libs\" \"--cflags\" \"openssl\"` did not exit successfully: exit code: 1\n--- stderr\nPackage openssl was not found in the pkg-config search path.\nPerhaps you should add the directory containing `openssl.pc\'\nto the PKG_CONFIG_PATH environment variable\nNo package \'openssl\' found\n"

  --- stderr
  thread 'main' panicked at '

  Could not find directory of OpenSSL installation, and this `-sys` crate cannot
  proceed without this knowledge. If OpenSSL is installed and this crate had
  trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
  compilation process.

  Make sure you also have the development packages of openssl installed.
  For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.

  If you're in a situation where you think the directory *should* be found
  automatically, please open a bug at https://github.com/sfackler/rust-openssl
  and include information about your system as well as this message.

  $HOST = x86_64-unknown-linux-gnu
  $TARGET = i686-unknown-linux-gnu
  openssl-sys = 0.9.58

  ', /cargo/registry/src/github.com-1ecc6299db9ec823/openssl-sys-0.9.58/build/find_normal.rs:157:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

rust-openssl can't find openssl config on Github Actions Ubuntu-latest even though i installed `libssl-dev` & `pkg-config` as stated in the documentation, did i miss something here?

my workflow looks like following

```yaml
on: [push, pull_request, release]

name: Build

jobs:
    build:
        name: Build
        runs-on: ubuntu-latest
        strategy:
            matrix:
                target:
                    - i686-pc-windows-gnu
                    - i686-unknown-linux-gnu
                    - x86_64-apple-darwin
                    - x86_64-pc-windows-gnu
                    - x86_64-unknown-linux-gnu
        steps:
            - uses: actions/checkout@v2
            - uses: actions-rs/toolchain@v1
              with:
                  toolchain: stable
                  target: ${{ matrix.target }}
                  override: true
            - name: Install Build Dependencies
              run: sudo apt-get update && sudo apt-get install pkg-config libssl-dev clang
            - uses: actions-rs/cargo@v1
              env:
                TARGET_CC: clang
              with:
                use-cross: true
                command: build
                args: --release --target=${{ matrix.target }}
```

**Comment by sfackler** [comment]
You are cross compiling. libssl-dev provides an x86_64 version of OpenSSL, but you need an x86 version.

**Comment by DevYukine** [comment]
Oh i see, i assume there is no package for that and i would need to build OpenSSL from source for that?

**Comment by sfackler** [comment]
There might be - I'm not super familiar with the options around cross compilation on Ubuntu. Regardless, you'd need to manually point openssl-sys to the right install with the environment variables rather than relying on pkg-config.

**Comment by DevYukine** [comment]
After searching for a while i found that there is a pre compiled package for `i386`, installed that and tried but for some reason for "x86_64-pc-windows-gnu" which is also x86_64 it still can't find the openssl installation, do you know if there is any better way than locally compile OpenSSH for every target system? it would be really annoying to do so in CI/CD

my current workflow attempt looks like this
```yaml
on: [push, pull_request, release]

name: Build

jobs:
    build:
        name: Build
        runs-on: ubuntu-latest
        strategy:
            matrix:
                target:
                    - i686-pc-windows-gnu
                    - i686-unknown-linux-gnu
                    - x86_64-apple-darwin
                    - x86_64-pc-windows-gnu
                    - x86_64-unknown-linux-gnu
        steps:
            - uses: actions/checkout@v2
            - uses: actions-rs/toolchain@v1
              with:
                  toolchain: stable
                  target: ${{ matrix.target }}
                  override: true
            - name: Install x86 Build Dependencies
              if: ${{ matrix.target == 'i686-pc-windows-gnu' || matrix.target == 'i686-unknown-linux-gnu' }}
              run: sudo dpkg --add-architecture i386 && sudo apt-get update && sudo apt-get install libssl-dev:i386 clang
            - name: Install x86_64 Build Dependencies
              if: ${{ matrix.target != 'i686-pc-windows-gnu' && matrix.target != 'i686-unknown-linux-gnu' }}
              run: sudo apt-get update && sudo apt-get install pkg-config libssl-dev clang
            - name: Build x86
              uses: actions-rs/cargo@v1
              if: ${{ matrix.target == 'i686-pc-windows-gnu' || matrix.target == 'i686-unknown-linux-gnu' }}
              env:
                  TARGET_CC: clang
                  OPENSSL_LIB_DIR: /usr/lib/i386-linux-gnu/
                  OPENSSL_INCLUDE_DIR: /usr/include/openssl/
              with:
                  command: build
                  args: --release --target=${{ matrix.target }}
            - name: Build x86_64
              uses: actions-rs/cargo@v1
              if: ${{ matrix.target != 'i686-pc-windows-gnu' && matrix.target != 'i686-unknown-linux-gnu' }}
              env:
                TARGET_CC: clang
              with:
                command: build
                args: --release --target=${{ matrix.target }}
```

**Comment by sfackler** [comment]
You can enable the vendored feature to have the crate build a local copy for you.

**Comment by DevYukine** [comment]
Thank you a lot! that works like a charm, cheers.

**Comment by dd6ds** [comment]
@DevYukine 

Hi,
tryed your yml above...thanks for this!
but fail in my case

with Ubuntu 24.04 /  Noble Numbat 

is there somthing to fix in the file for Ubuntu 24.04 ?


<img width="472" height="395" alt="Image" src="https://github.com/user-attachments/assets/37aa4e8e-ff81-4a75-b5e7-ac4947b5ac61" />

<img width="1335" height="674" alt="Image" src="https://github.com/user-attachments/assets/028d8e97-61d3-4ca0-819b-e7066573fed3" />


<img width="633" height="711" alt="Image" src="https://github.com/user-attachments/assets/f0c7312c-4b89-466e-8bd1-2e5fe28d4f37" />

**Comment by dd6ds** [comment]
when reduce the targets then it works for windowns and linux x86_64

```
                target:
                    - x86_64-unknown-linux-gnu
#                    - i686-unknown-linux-gnu
#                    - i686-pc-windows-gnu
                    - x86_64-pc-windows-gnu
#                    - x86_64-apple-darwin

```
